code/espurna/settings_embedis.h

```diff
@@ -0,0 +1,674 @@
+                break;
+            }
+            callback(std::move(kv));
+        } while (_state != State::End);
+    }
+
+    // read every key into a vector
+    std::vector<String> keys() {
+        std::vector<String> out;
+        out.reserve(count());
+
+        foreach([&](KeyValueResult&& kv) {
+            out.push_back(kv.key.read());
+        });
+
+        return out;
+    }
+
+    // set or update key with value contents. ensure 'key' isn't empty, 'value' can be empty
+    bool set(const String& key, const String& value) {
+
+        // ref. 'estimate()' implementation in regards to the storage calculation
+        auto need = estimate(key, value);
+        if (!need) {
+            return false;
+        }
+
+        auto key_len = key.length();
+        auto value_len = value.length();
+
+        Cursor to_erase(_storage);
+        bool need_erase = false;
+
+        auto start_pos = _cursor_reset_end();
+
+        do {
+            auto kv = _read_kv();
+            if (!kv) {
+                break;
+            }
+
+            start_pos = kv.value.cursor.begin;
+
+            // in the very special case we can match the existing key
+            if ((kv.key.length == key_len) && (kv.key.read() == key)) {
+                if (kv.value.length == value.length()) {
+                    if (kv.value.read() == value) {
+                        return true;
+                    }
+                    start_pos = kv.key.cursor.end;
+                    break;
+                }
+                // but we may need to write over it, when contents are different
+                to_erase.reset(kv.value.cursor.begin, kv.key.cursor.end);
+                need_erase = true;
+            }
+
+        } while (_state != State::End);
+
+        if (need_erase) {
+            _raw_erase(start_pos, to_erase);
+            start_pos += to_erase.size();
+        }
+
+        // we should only insert when possition is still within possible size
+        if (start_pos && (start_pos >= need)) {
+            auto writer = Cursor::fromEnd(_storage, start_pos - need, start_pos);
+
+            // put the length of the value as 2 bytes and then write the data
+            (--writer).write(key_len & 0xff);
+            (--writer).write((key_len >> 8) & 0xff);
+            while (key_len--) {
+                (--writer).write(key[key_len]);
+            }
+
+            (--writer).write(value_len & 0xff);
+            (--writer).write((value_len >> 8) & 0xff);
+
+            if (value_len) {
+                while (value_len--) {
+                    (--writer).write(value[value_len]);
+                }
+            } else {
+                (--writer).write(0);
+                (--writer).write(0);
+            }
+
+            // we also need to pad the space *after* the value
+            // but, only when we still have some space left
+            if ((start_pos - need) >= 2) {
+                _cursor_set_position(writer.begin);
```

Comment:
Ah, I thought this would be harder to find. Never-mind the previous workaround comment, can you please test this instead?
```suggestion
                _cursor_set_position(writer.begin - _cursor.begin);
```
Suggested 'test' still stands, key should not disappear.